### ISSUE TITLE
secure thread starvation

### ISSUE THREAD

**Issue by vtcao297** [OP-TEE/optee_os#4438]
Hi,

On our platform we have a secure watchdog service within optee that exposes an interface which a Linux normal world daemon would regularly send keep-alive request to every 1min to refresh the timer via the OPTEE client app interface.  Our SoC is based on Cortex-A53 with 4 cores so OPTEE is configured for 4 threads.  Things are working okay if aren't an overloaded number of requests to OPTEE.

However, we are seeing a situation if there are many requests into optee like when we have a service to encrypt large chunks of data that the secure wdt daemon keep-alive invoke request to fail to get serviced since eventually cause the system to reboot because the timer didn't get refresh.  What we see if that the wdt daemon request is getting put into the wait queue in the optee kernel driver because no threads we're available.  But even when the encryption activity is completed, the queued daemon request into TEE doesn't seem to be serviced (it just seems to wait indefinitely).  If we kill the wdt daemon service and restart it then the request is serviced again successfully.

We are able to reproduce this pretty consistently when running the full xtest suite.  Also what seems to help a bit is by introducing a slight delay in between each invoke request to the OPTEE, but this doesn't seem like a good option as it would introduce some performance hit so I would like some help to understand why the queue request aren't being service after the thread resource become available?

**Comment by jenswi-linaro** [comment]
The queued requests are supposed to be served as soon as a thread becomes available. Perhaps there's an error in the logic?

**Comment by jenswi-linaro** [comment]
> I'm been trying to debug this issue the past 2 weeks and not getting any closer to root-cause why this happens. If we hit this issue one of the process (the wdt daemon) will go into a uninterruptible sleep (Ds state from px -aux) so i have a few questions:
> 
> 1.) What is the expectation when you have 2 client applications calling two separate TAs at the **same time** and only 1 Secure Thread is free/avialable? For example, let say CA1 calls TEECInvokeCommand into PTA1 interface and another CA2 does a TEECOpenSession followed by TEECInvokeCommand into UTA1.

CA1 will get a thread in secure world to serve the invoke command. CA2 will hang in `optee_cq_wait_for_completion()` waiting for someone to call `optee_cq_wait_final()`.

The task of CA1 is preempted in Linux it will exit secure world via a non-secure interrupt with it will be served in Linux probably before `optee_handle_rpc()` has been called. Eventually CA1 will be granted CPU time again and can reenter secure world with `OPTEE_SMC_CALL_RETURN_FROM_RPC` filled in by `optee_handle_rpc()`. When the invoke command is complete OP-TEE frees the secure thread and returns to normal world. There the task of CA1 will break out of the `while (true)` loop and call `optee_cq_wait_final()` which will wake up CA2.

If CA2 is unlucky yet another CA may claim the secure world thread and cause CA2 to wait in `optee_cq_wait_for_completion()` once more.

> 
> 2.) If CA1 is able to do a SMC request and gets assigned a secure thread in thread_alloc_and_run() to do the invokecommand, this is atomic and uninterrupted until the invokecommand returns? If it's not atomic and it is interrupted, how would the linux kernel know to reschedule a optee_os thread?

The call is not atomic in the sense that it can return for RPC processing, including rescheduling.
The Linux kernel will continue to schedule the task of CA1 as if it was executing inside the kernel.

> 
> 3.) Say if from normal world let say I run xtest and I Ctrl-C the xtest program in the middle of execution, how does optee os know to free up the secure thread that was assigned to xtest?

The sessions are cleaned up when a process is destroyed so the cleanup function in the driver is supposed to take care of that.

> 
> I enabled all kind of debug traces logs i can think (loglevel 4, stack check, lockdep, etc) but it hasn't been helpful to narrow down where the problem area is that is causing this issue. I think I mentioned it before but if start-up another process instance of the watchdog daemon all the request would be serviced again, but of course if another CA happens to trigger a command into OPTEE at the same time again (say we run xtest again as an example) the issue will happen again.

Hmm, maybe CA1 manages to get killed in a way that `optee_cq_wait_final()` isn't called?

**Comment by divneil** [comment]
@jenswi-linaro I have started looking at this issue recently. Here are my observations till now, I am hoping that it will help in finding the the root-cause.

The WDT CA is stuck in response to the `OPTEE_RPC_WAIT_QUEUE_SLEEP` command sent by the optee-os kernel.
We know that this command is sent by the `mutex_lock()` being executed in the flow.
The WDT PTA does not have `TA_FLAG_CONCURRENT` set, and it gets stuck in `lock_single_instance()` after other TAs/PTAs start executing.

On setting `TA_FLAG_CONCURRENT` for this PTA, I have not seen WDT CA getting stuck.

```
static bool tee_ta_try_set_busy(struct tee_ta_ctx *ctx)
{
        if (ctx->flags & TA_FLAG_CONCURRENT)
                return true; /* I used TA_FLAG_CONCURRENT to make sure it returns from here */
        ......

        if (ctx->flags & TA_FLAG_SINGLE_INSTANCE) {
                lock_single_instance();
        }
        ......
}
```

**Comment by jenswi-linaro** [comment]
Perhaps this is fixed by https://github.com/OP-TEE/optee_os/pull/4514

**Comment by divneil** [comment]
> Perhaps this is fixed by #4514

Yeah! we are testing with this patch, all seems to be well for a day now. @vtcao297 or myself will keep you posted.

**Comment by vtcao297** [comment]
Closing as this issue is now fixed and tested.
